### PR TITLE
adding to_datacite method

### PR DESCRIPTION
**Pull request by djarecka**
this is a continuation of #516 (I had issues with using the secrets, so testing this from the dandi branch)

Adding a new method `to_dandicite` that creates a datacite dictionary from the metadata object

### CODE REVIEW

**Comment by djarecka** [comment]
@satra -  I'm testing metadata that I took from dandisets 04 and 08. I could modify dataset 27 and add to the test, but can also keep adding various metadata to [test_dantimeta_datacite](https://github.com/dandi/dandi-cli/pull/596/files#diff-4b008975a2df2bb2ddf1030178e22a3dc3585fabaa159d35f9b74d1b394acdf9R306)

**Comment by satra** [comment]
for the moment you can use a local test, but we should also add a test which downloads from the API, augments it with the missing pieces (because draft metadata may not have doi, url, etc.,.) and then runs through your datacite test using the test server.

File: dandi/models.py
Comment (satra) [inline dandi/models.py:1059]: this seems off and does not seem to take into account the inheritance in the current model.

File: dandi/tests/test_models.py
Comment (satra) [inline dandi/tests/test_models.py:190]: why activities and not compare to the posted metadata?

Comment (satra) [reply dandi/tests/test_models.py:190]: also we shouldn't have to do the `doi.replace`

File: dandi/tests/test_models.py
Comment (satra) [inline dandi/tests/test_models.py:201]: also should not need this.

File: dandi/tests/test_models.py
Comment (satra) [inline dandi/tests/test_models.py:328]: i think i include datetime in datePublished in the model, and should be in isoformat with timezone info.

File: dandi/models.py
Comment (djarecka) [reply dandi/models.py:1059]: inheriting from `DandisetMeta` seems reasonable to me, please let me know what you would like to change

Comment (satra) [reply dandi/models.py:1059]: please see the current model. i think you changed with merging master.

File: dandi/tests/test_models.py
Comment (djarecka) [reply dandi/tests/test_models.py:190]: indeed, the replacement is not needed. But the form with "activities" I took from datacite examples and it doesn't work without

File: dandi/tests/test_models.py
Comment (djarecka) [reply dandi/tests/test_models.py:328]: so I should be using `Publishable` class or some version of `DandisetMeta`, e.g. `PublishedDandisetMeta`?

File: dandi/models.py
Comment (djarecka) [reply dandi/models.py:1059]: oh, i see, the merge was not perfect..

**Comment by djarecka** [comment]
> for the moment you can use a local test, but we should also add a test which downloads from the API, augments it with the missing pieces (because draft metadata may not have doi, url, etc.,.) and then runs through your datacite test using the test server.

And do we have any good way of getting metadata without downloading datasets? If we want to tests existing datasets starting from API, I'd have to somehow get the metadata without downloading the content, and not sure what is the best way.

Right now, I use datalad to get repo (without the content) and then I use `Dandiset(path_to_repo)`.

**Comment by yarikoptic** [comment]
Yes, api. https://github.com/dandi/dandi-cli/blob/master/dandi/dandiapi.py#L301 provides python interface

**Comment by satra** [comment]
right now dataset metadata is embedded in the returned json.

**Comment by djarecka** [comment]
@satra @yarikoptic - thanks! I've used `DandiAPIClient("https://api.dandiarchive.org/api/")` without any tokens, and `version=draft`. Let me know if I should use something else

**Comment by djarecka** [comment]
@satra @yarikoptic - please let me know if you have any suggestions for the changes or the tests

File: dandi/tests/test_models.py
Comment (yarikoptic) [reply dandi/tests/test_models.py:398]: attn @djarecka

File: dandi/tests/test_models.py
Comment (yarikoptic) [reply dandi/tests/test_models.py:201]: may be there is some alternative route as well, e.g. - may be they have a schema (or some validator) to validate against without pestering API?

Comment (satra) [reply dandi/tests/test_models.py:201]: only their xml schema is up to standard. their json schema is some bits of work in progress. i think a CI test is fine. we will also do this as part of the api validation check.

File: dandi/tests/test_models.py
Comment (yarikoptic) [inline dandi/tests/test_models.py:216]: I think we could use the one we have in the fixture here without relying on the live one and existence of specific dandisets there.

only 000027 (well and 000027) is the dandiset we have under "our control" and the ones which could be used in the tests.

if tests talk to real server, I think we should annotate them somehow, probably with `@pytest.mark.integration`.

@jwodder please review and provide comments and suggestions to get this PR landed.

File: dandi/models.py
Comment (jwodder) [inline dandi/models.py:185]: Why is this called `publicationYear` when it's being set to a timestamp?  Is that intentional?

File: dandi/models.py
Comment (jwodder) [inline dandi/models.py:116]: As this is only used for checking containment, it may be more efficient to make it a set.

File: dandi/models.py
Comment (jwodder) [inline dandi/models.py:274]: You should probably raise something more specific than an `Exception`, like a `ValueError`.

File: dandi/tests/test_models.py
Comment (jwodder) [inline dandi/tests/test_models.py:220]: This test should be skipped if `DATACITE_DEV_PASSWORD` is not set so that the test suite can be run without needing that secret.

File: dandi/tests/test_models.py
Comment (jwodder) [inline dandi/tests/test_models.py:206]: You should call `sr.raise_for_status()` to ensure that the request was successful.

File: dandi/tests/test_models.py
Comment (jwodder) [inline dandi/tests/test_models.py:198]: You should call `.raise_for_status()` to ensure that the request is successful.

File: dandi/tests/test_models.py
Comment (jwodder) [inline dandi/tests/test_models.py:269]: This does not do what you think it does.  It simply tests that the list is nonempty.  Presumably you meant to wrap this in `all()`?

File: dandi/tests/test_models.py
Comment (jwodder) [inline dandi/tests/test_models.py:266]: Below, you seem to be assuming that the exception is a pydantic `ValidationError`.  This should thus be changed to `with pytest.raises(ValidationError) as exc:`.

File: dandi/tests/test_models.py
Comment (jwodder) [inline dandi/tests/test_models.py:373]: This test should be skipped if `DATACITE_DEV_PASSWORD` is not set so that the test suite can be run without needing that secret.

File: dandi/tests/test_models.py
Comment (jwodder) [inline dandi/tests/test_models.py:193]: Is there a reason to assert on the status codes instead of calling `raise_for_status()` on the responses?

File: dandi/tests/test_models.py
Comment (jwodder) [inline dandi/tests/test_models.py:25]: This appears to be unused.

**Review by jwodder — CHANGES_REQUESTED** [review]

File: dandi/models.py
Comment (djarecka) [reply dandi/models.py:116]: yes, can move it to set

File: dandi/tests/test_models.py
Comment (djarecka) [reply dandi/tests/test_models.py:269]: yes, i did mean to add run `all`, thanks!

File: dandi/tests/test_models.py
Comment (djarecka) [reply dandi/tests/test_models.py:198]: this is sometimes run "just in case" before creating a new doi, so it can return an error

File: dandi/models.py
Comment (djarecka) [reply dandi/models.py]: moved to `meta.datePublished.year`

File: dandi/tests/test_models.py
Comment (djarecka) [reply dandi/tests/test_models.py]: probably not, changed it

File: dandi/tests/test_models.py
Comment (djarecka) [reply dandi/tests/test_models.py:399]: @satra - sorry, forgot to answer, but looks like `str(datetime.now().year)` works fine - do you expect that the error should be raised?

File: dandi/tests/test_models.py
Comment (djarecka) [reply dandi/tests/test_models.py:213]: I also could come back to the previous version where I used metadata from real dandiset as examples, but I saved the metadata as a json file in the repo

**Comment by yarikoptic** [comment]
Please merge master or rebase -- "Test populate_dandiset_yaml.py / test (push) " should be gone now.

**Comment by yarikoptic** [comment]
didn't look through all details (we should put con/tinuous to work here, filed #639) but those failing which are saw are of the 
```
dandi/tests/test_models.py:232: in test_datacite
    newmeta = PublishedDandisetMeta(**newmeta_js)
pydantic/main.py:406: in pydantic.main.BaseModel.__init__
    ???
E   pydantic.error_wrappers.ValidationError: 1 validation error for PublishedDandisetMeta
E   assetsSummary
E     field required (type=value_error.missing)
```
nature

**Comment by djarecka** [comment]
the biggest problem is that `assetsSummary` was added to `PublishedDandisetMeta`. Do you have any function to create it?

**Comment by satra** [comment]
that would be a good function to right. this function should iterate over all assets in a dandiset to create a summary of unique values for those fields.

however, this is also what yarik was going to add to the asset. for example approach/technique are not being populated at the asset level right now. we can relax some of the components to optional, but i think species, approach, technique should really be there.

**Comment by djarecka** [comment]
i'm a bit confused since it looks like a combination of information that should be provided by the author and info that could be calculated automatically...

**Comment by satra** [comment]
all of that info is calculated from asset metadata, hence the name. the authors would provide this when creating nwb files (and they generally do). we have to do a bit more magic for bids datasets, so those would likely not be publishable till we have figured out some pieces.

**Comment by djarecka** [comment]
ok, so for testing, I guess it's ok for me to use some random entry

**Comment by satra** [comment]
> I guess it's ok for me to use some random entry

yes

**Comment by djarecka** [comment]
for now, I just saved meta from 000004 and 00008 in `tests/data/metadata` for testing. I could work on dandiset 000027 and additional dandisets later. Today I was not really able to use `https://gui-staging.dandiarchive.org/` (doesn't load) and I still can't edit `000027`

**Comment by djarecka** [comment]
not sure what to do with `Test model schemata / publish` test

**Comment by yarikoptic** [comment]
well, if you look at that fail run fail you will see
```
diff --git a/releases/0.3.1/published-dandiset.json b/releases/0.3.1/published-dandiset.json
index 14b909b..3f03ea5 100644
--- a/releases/0.3.1/published-dandiset.json
+++ b/releases/0.3.1/published-dandiset.json
@@ -270,7 +270,8 @@
     "contributor",
     "license",
     "assetsSummary",
-    "version"
+    "version",
+    "doi"
   ],
   "definitions": {
     "RoleType": {
Error:  Existing schema files modified instead of creating a new version
Error: Process completed with exit code 1.
```
so this PR modifies schema by adding `doi` (I am just surprised we did not have it for published-dandiset yet).  I guess it is intended, so let's just boost schema version to 0.3.2 and merge.  Or @satra -- you want to mutate 0.3.1? change is minor but not sure if would not cause some ripples in dandi-api / webui.

**Comment by satra** [comment]
@yarikoptic - the doi was already there in the model, and should also be non-optional. this mutates it by removing a default value, which i think has an impact on whether or not it got exported to JSON schema. so this was a bug in the schema generation which this PR fixes.

File: dandi/tests/test_models.py
Comment (yarikoptic) [reply dandi/tests/test_models.py:195]: shouldn't it matter what kind of error it is?  e.g. if we disconnected an not removed DOI -- we should try again.  `grep` for `tenacity` to see how we orchestrate retries ATM.

Note: we now have 57 DOIs in test. instance of datacite, some of them seems to be for the same dataset(s).  I wonder if some of those are due to not carrying if cleanup succeeded or not, and shouldn't we actually cleanup most if not all of them? ;-)

File: dandi/models.py
Comment (yarikoptic) [inline dandi/models.py:282]: seems to be also of `ValueError` kind and we better inform the value on which we choked:

```suggestion
                raise ValueError(f"identifier is expected to be type:number, got {rel_el.identifier}")
```

**Comment by yarikoptic** [comment]
ok, left minor comments but could be handled post-merge/migration to dedicated `dandischema`. 
So I guess we could just wait for @dchiquito to give blessing to mutate or could boost the .patch and proceed with the merge.

File: dandi/models.py
Comment (yarikoptic) [reply dandi/models.py:282]: Let  me just proceed with this invaluable contribution of mine -- hopefully it doesn't break anything

File: dandi/tests/test_models.py
Comment (yarikoptic) [reply dandi/tests/test_models.py]: yeah -- we should not rely on testing on real arbitrary datasets out of our control, at least until we know that they are metadata records are as kosher as they could be and that we have safeguards that nobody could make them invalid ;)   but ok, let's proceed as is for now and adjust later.

Comment (djarecka) [reply dandi/tests/test_models.py]: this has been removed after our discussion on Monday

**Comment by yarikoptic** [comment]
Ok, @dchiquito gave blessing in slack, I fixed lint issue I introduced, let's proceed.  Anything remaining should be done against `dandischema`

File: dandi/tests/test_models.py
Comment (yarikoptic) [reply dandi/tests/test_models.py]: Thank you @djarecka ! I must had been looking at the outdated diff while checking on this before merge :-/ Great!

**Pull request merged**
